infra/src/main/kotlin/fi/fta/geoviite/infra/linking/switches/SwitchLinkingService.kt

```diff
@@ -0,0 +1,847 @@
+                )
+            }
+        }.sortedBy { it.m }
+    }
+}
+
+private fun findExistingSwitchEdgeSegmentWithSwitchFreeAdjacentSegment(
+    existingSwitchId: IntId<TrackLayoutSwitch>,
+    layoutSegments: List<LayoutSegment>,
+    searchIndexRange: IntProgression,
+): IndexedValue<LayoutSegment>? {
+    val layoutSegmentIndicesAreValid =
+        searchIndexRange.first in layoutSegments.indices && searchIndexRange.last in layoutSegments.indices
+
+    val step = searchIndexRange.step
+    val firstAdjacentIndexIsValid = (searchIndexRange.first + step) in layoutSegments.indices
+    val lastAdjacentIndexIsValid = (searchIndexRange.last + step) in layoutSegments.indices
+
+    val adjacentSegmentIndicesAreValid = firstAdjacentIndexIsValid && lastAdjacentIndexIsValid
+
+    require(layoutSegmentIndicesAreValid) {
+        "Invalid searchIndexRange: $searchIndexRange contains indices outside of layoutSegments (${layoutSegments.indices})"
+    }
+
+    require(adjacentSegmentIndicesAreValid) {
+        "Invalid searchIndexRange: $searchIndexRange contains adjacent indices outside of layoutSegments (${layoutSegments.indices})"
+    }
+
+    for (i in searchIndexRange) {
+        val segment = layoutSegments[i];
+
+        val existingSwitchIdMatchesSegment = existingSwitchId == segment.switchId
+        if (!existingSwitchIdMatchesSegment) {
+            return null
+        }
+
+        val adjacentSegmentHasNoSwitch = layoutSegments[i + searchIndexRange.step].switchId == null
+        if (adjacentSegmentHasNoSwitch) {
+            return IndexedValue(i, segment)
+        }
+    }
+
+    return null
+}
+
+fun tryToSnapOverlappingSwitchSegmentToNearbySegment(
+    layoutAlignment: LayoutAlignment,
+    match: FittedSwitchJointMatch,
+): FittedSwitchJointMatch {
+    val referencedLayoutSegment = layoutAlignment.segments[match.segmentIndex]
+
+    val segmentIsSwitchFree = referencedLayoutSegment.switchId == null
+    if (segmentIsSwitchFree) {
+        return match
+    }
+
+    // Snapping towards the start of the location track.
+    match.segmentIndex
+        .takeIf { segmentIndex -> segmentIndex > 1 }
+        ?.let { segmentIndex -> IntProgression.fromClosedRange(segmentIndex, 1, -1) }
+        ?.let { negativeSearchIndexDirection ->
+            findExistingSwitchEdgeSegmentWithSwitchFreeAdjacentSegment(
+                referencedLayoutSegment.switchId as IntId,
+                layoutAlignment.segments,
+                searchIndexRange = negativeSearchIndexDirection,
+            )
+        }
+        ?.let { indexedExistingSwitchStartSegment ->
+            val distanceToPreviousSwitchLineStart =
+                match.m - indexedExistingSwitchStartSegment.value.startM
+            val hasAdjacentLayoutSegment = indexedExistingSwitchStartSegment.index > 0
+
+            if (hasAdjacentLayoutSegment && distanceToPreviousSwitchLineStart <= MAX_SWITCH_JOINT_OVERLAP_CORRECTION_AMOUNT_METERS) {
+                return match.copy(
```

Comment:
Nämäkin lettien sisällä olevat returnit on ainakin minulle aina vähän hämmentäviä. Tässäkin pitää huomioida sekä nuo nullablet että if jotta tietää palauttaako tämä ja lisäksi muistaa että tuo return palauttaa koko funkkarista vs vain `match.copy` palauttaisi letistä.

Tuon voisi varmaan pilkkoa esim niin että tekisi erilliset funkkarit `findMatchTowardsStartOfTrack` ja `findMatchTowardsEndOfTrack` ja sitten täällä ylätasolla
```
return findMatchTowardsStartOfTrack(..)
 ?: findMatchTowardsEndOfTrack(..)
 ?: match
 ```

Reply:
Sinällään nuo eteen- ja taaksepäin haut näyttää niin läheisiltä että siinä olisi ehkä joku yhteinenkin abstraktio vielä löydettävissä. Tuo indeksipuljaus lienee ainoa ero.

Reply:
Jälleen meikäläisen koodia. Indeksipuljaus on tosiaan ainut ero, mutta ero on nähtävissä eksplisiittisesti vrt. raskaampaan päättelyyn esim. listasta peräisin olevilla indeksieroilla ja vertailulambdalla. Tulkitsin aiemmin luettavammaksi tässä formaatissa.

Funktiointi voisi silti tosiaan olla jees.